From 79826a22f9a9bb0223734293dd382435ce238d0c Mon Sep 17 00:00:00 2001
From: Ilya Ledvich <ilya@compulab.co.il>
Date: Thu, 11 Aug 2016 15:27:10 +0300
Subject: [PATCH 42/46] ARM: dts: imx7d: cl-som-imx7: change codec clock source

Change clock source for analog audio codec wm8731 from 12MHz
external crystal to 12.288MHz audio master clock (MCLK).
The MCLK is generated by SAI1 interface and sourced to the
codec from GPIO1_IO01 pin muxed in SAI1_MCLK mode.

Signed-off-by: Ilya Ledvich <ilya@compulab.co.il>
---
 arch/arm/boot/dts/imx7d-cl-som-imx7.dts | 21 ++++++++++++++-------
 arch/arm/mach-imx/clk-imx7d.c           |  2 +-
 2 files changed, 15 insertions(+), 8 deletions(-)

diff --git a/arch/arm/boot/dts/imx7d-cl-som-imx7.dts b/arch/arm/boot/dts/imx7d-cl-som-imx7.dts
index 52c40ed..01510d5 100644
--- a/arch/arm/boot/dts/imx7d-cl-som-imx7.dts
+++ b/arch/arm/boot/dts/imx7d-cl-som-imx7.dts
@@ -125,23 +125,28 @@
 		simple-audio-card,frame-master = <&sound_master>;
 		/*simple-audio-card,bitclock-inversion;*/
 
+		assigned-clocks = <&clks IMX7D_AUDIO_MCLK_ROOT_SRC>,
+			          <&clks IMX7D_AUDIO_MCLK_ROOT_CLK>;
+		assigned-clock-parents = <&clks IMX7D_PLL_AUDIO_POST_DIV>;
+		assigned-clock-rates = <0>, <12288000>;
+
 		sound_master: simple-audio-card,cpu {
 			sound-dai = <&sai1 0>;
-			system-clock-direction = "in";
+			system-clock-direction = "out";
 		};
 
 		simple-audio-card,codec {
 			sound-dai = <&wm8731>;
-			system-clock-type = "xtal";
-			system-clock-frequency = <12000000>;
-			system-clock-direction = "out";
+			system-clock-type = "mclk";
+			system-clock-frequency = <12288000>;
+			system-clock-direction = "in";
 		};
 	};
 };
 
 &clks {
 	assigned-clocks = <&clks IMX7D_PLL_AUDIO_POST_DIV>;
-	assigned-clock-rates = <786432000>;
+	assigned-clock-rates = <884736000>;
 };
 
 &cpu0 {
@@ -439,7 +444,7 @@
 
 		pinctrl_sai1_lpsr: sai1grp_lpsr {
 			fsl,pins = <
-				MX7D_PAD_GPIO1_IO01__SAI1_MCLK		0x14
+				MX7D_PAD_GPIO1_IO01__SAI1_MCLK			0x14
 			>;
 		};
 
@@ -552,6 +557,8 @@
 		#sound-dai-cells = <0>;
 		compatible = "wlf,wm8731";
 		reg = <0x1a>;
+		clocks = <&clks IMX7D_AUDIO_MCLK_ROOT_CLK>;
+		clock-names = "mclk";
 		status = "okay";
 	};
 
@@ -581,7 +588,7 @@
 	assigned-clocks = <&clks IMX7D_SAI1_ROOT_SRC>,
 			  <&clks IMX7D_SAI1_ROOT_CLK>;
 	assigned-clock-parents = <&clks IMX7D_PLL_AUDIO_POST_DIV>;
-	assigned-clock-rates = <0>, <12288000>;
+	assigned-clock-rates = <0>, <36864000>;
 	status = "okay";
 };
 
diff --git a/arch/arm/mach-imx/clk-imx7d.c b/arch/arm/mach-imx/clk-imx7d.c
index 980a30a..cbda294 100644
--- a/arch/arm/mach-imx/clk-imx7d.c
+++ b/arch/arm/mach-imx/clk-imx7d.c
@@ -856,7 +856,7 @@ static void __init imx7d_clocks_init(struct device_node *ccm_node)
 	clks[IMX7D_WDOG3_ROOT_CLK] = imx_clk_gate4("wdog3_root_clk", "wdog_post_div", base + 0x49e0, 0);
 	clks[IMX7D_WDOG4_ROOT_CLK] = imx_clk_gate4("wdog4_root_clk", "wdog_post_div", base + 0x49f0, 0);
 	clks[IMX7D_CSI_MCLK_ROOT_CLK] = imx_clk_gate4("csi_mclk_root_clk", "csi_mclk_post_div", base + 0x4490, 0);
-	clks[IMX7D_AUDIO_MCLK_ROOT_CLK] = imx_clk_gate4("audio_mclk_root_clk", "audio_mclk_post_div", base + 0x4790, 0);
+	clks[IMX7D_AUDIO_MCLK_ROOT_CLK] = imx_clk_gate2_flags("audio_mclk_root_clk", "audio_mclk_post_div", base + 0x4790, 0, CLK_SET_RATE_PARENT);
 	clks[IMX7D_WRCLK_ROOT_CLK] = imx_clk_gate4("wrclk_root_clk", "wrclk_post_div", base + 0x47a0, 0);
 	clks[IMX7D_USB_CTRL_CLK] = imx_clk_gate4("usb_ctrl_clk", "ahb_root_clk", base + 0x4680, 0);
 	clks[IMX7D_USB_PHY1_CLK] = imx_clk_gate4("usb_phy1_clk", "pll_usb1_main_clk", base + 0x46a0, 0);
-- 
1.9.1

